File: HelloWorld/app/src/main/java/com/banshion/helloworld/MainActivity.java
package com.banshion.helloworld;

import androidx.appcompat.app.AppCompatActivity;

import android.content.Intent;
import android.os.Bundle;
import android.view.View;
import android.widget.Button;

import com.banshion.helloworld.datastorage.DataStorageActivity;

public class MainActivity extends AppCompatActivity implements View.OnClickListener {

    private Button m_UIBtn, m_EventBtn,m_DataStorage;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main2);

        m_UIBtn = findViewById(R.id.ui);
        m_EventBtn = findViewById(R.id.event);
        m_DataStorage = findViewById(R.id.data_storage);
        m_UIBtn.setOnClickListener(this);
        m_EventBtn.setOnClickListener(this);
        m_DataStorage.setOnClickListener(this);
    }

    @Override
    public void onClick(View v) {
        Intent intent = null;
        switch (v.getId()) {
            case R.id.ui:
                intent = new Intent(MainActivity.this, UIActivity.class);
                break;
            case R.id.event:
                intent = new Intent(MainActivity.this, EventActivity.class);
                break;
            case R.id.data_storage:
                intent = new Intent(MainActivity.this, DataStorageActivity.class);
                break;
        }
        startActivity(intent);
    }


}

File: RuntimePermissionTest/app/src/main/java/com/banshion/runtimepermissiontest/MainActivity.kt
package com.banshion.runtimepermissiontest

import android.Manifest
import android.content.Intent
import android.content.pm.PackageManager
import android.net.Uri
import androidx.appcompat.app.AppCompatActivity
import android.os.Bundle
import android.widget.Toast
import androidx.core.app.ActivityCompat
import androidx.core.content.ContextCompat
import androidx.core.content.ContextCompat.startActivity
import kotlinx.android.synthetic.main.activity_main.*

class MainActivity : AppCompatActivity() {
    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_main)
        makeCall.setOnClickListener{
            //判断是否授权
            if (ContextCompat.checkSelfPermission(this,Manifest.permission.CALL_PHONE) != PackageManager.PERMISSION_GRANTED) {
                // 请求授权
                // 第一个参数要求是Activity的实例；
                // 第二个参数是一个String数组，我们把要申请的权限名放在数组中即可；
                // 第三个参数是请求码，只要是唯一值就可以了
                ActivityCompat.requestPermissions(this,arrayOf(Manifest.permission.CALL_PHONE), 1)
            } else {
                call()
            }
        }
    }

    /***
     * 调用完requestPermissions()方法之后，
     * 系统会弹出一个权限申请的对话框，用户可以选择同意或拒绝我们的权限申请。
     * 不论是哪种结果，最终都会回调到onRequestPermissionsResult()方法中，而授权的结果则会封装在grantResults参数当中。
     */
    override fun onRequestPermissionsResult(requestCode: Int, permissions: Array<out String>, grantResults: IntArray) {
        super.onRequestPermissionsResult(requestCode, permissions, grantResults)
        when (requestCode) {
            1 -> {
                if (grantResults.isNotEmpty() && grantResults[0] == PackageManager.PERMISSION_GRANTED) {
                    call()
                } else {
                    Toast.makeText(this, "You denied the permission",
                            Toast.LENGTH_SHORT).show()
                }
            }
        }
    }

    private fun call() {
        try {
            val intent = Intent(Intent.ACTION_CALL)
            intent.data = Uri.parse("tel:10086")
            startActivity(intent)
        } catch (e: SecurityException) {
            e.printStackTrace()
        }
    }

}

